bpf/unwinders/go_traceid.h

```diff
@@ -0,0 +1,210 @@
+{
+	for (int i = 0; i < size; i++)
+	{
+		if (s1[i] != s2[i])
+		{
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static __always_inline void hex_string_to_bytes(char *str, u32 size, unsigned char *out)
+{
+	for (int i = 0; i < (size / 2); i++)
+	{
+		char ch0 = str[2 * i];
+		char ch1 = str[2 * i + 1];
+		u8 nib0 = (ch0 & 0xF) + (ch0 >> 6) | ((ch0 >> 3) & 0x8);
+		u8 nib1 = (ch1 & 0xF) + (ch1 >> 6) | ((ch1 >> 3) & 0x8);
+		out[i] = (nib0 << 4) | nib1;
+	}
+}
+
+static __always_inline bool get_trace_id(unsigned char *res_trace_id) {
+    long res;
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+    if (task == NULL) {
+        return false;
+    }
+
+    // It appears from all Go binaries we looked at 0xfffffffffffffff8 is the offset of `runtime.g`.
+    u64 g_addr_offset = 0xfffffffffffffff8;
+    
+    size_t g_addr;
+    res = bpf_probe_read_user(&g_addr, sizeof(void *), (void*)(BPF_CORE_READ(task, thread.fsbase)+g_addr_offset));
+    if (res < 0)
+    {
+        return false;
+    }
+    
+    // DW_TAG_member
+    //   DW_AT_name    ("m")
+    //   DW_AT_data_member_location    (48)
+    //   DW_AT_type    (0x0000000000088e39 "runtime.m *")
+    //   DW_AT_GO_embedded_field       (0x00)
+    size_t m_ptr_addr;
+    res = bpf_probe_read_user(&m_ptr_addr, sizeof(void *), (void*)(g_addr+48));
+    if (res < 0)
+    {
+        return false;
+    }
+    
+    // DW_TAG_member
+    //   DW_AT_name    ("curg")
+    //   DW_AT_data_member_location    (192)
+    //   DW_AT_type    (0x00000000000892b1 "runtime.g *")
+    //   DW_AT_GO_embedded_field       (0x00)
+    size_t curg_ptr_addr;
+    res = bpf_probe_read_user(&curg_ptr_addr, sizeof(void *), (void*)(m_ptr_addr+192));
+    if (res < 0)
+    {
+        return false;
+    }
+
+    // DW_TAG_member
+    //   DW_AT_name    ("labels")
+    //   DW_AT_data_member_location    (360)
+    //   DW_AT_type    (0x000000000005c242 "void *")
+    //   DW_AT_GO_embedded_field       (0x00)
+    void *labels_map_ptr_ptr;
+    res = bpf_probe_read_user(&labels_map_ptr_ptr, sizeof(void *), (void*)(curg_ptr_addr+360));
+    if (res < 0)
+    {
+        return false;
+    }
+
+    void *labels_map_ptr;
+    res = bpf_probe_read(&labels_map_ptr, sizeof(labels_map_ptr), labels_map_ptr_ptr);
+    if (res < 0)
+    {
+        return false;
+    }
+
+    u64 labels_count = 0;
+    res = bpf_probe_read(&labels_count, sizeof(labels_count), labels_map_ptr);
+    if (res < 0)
+    {
+        return false;
+    }
+    if (labels_count == 0)
+    {
+        return false;
+    }
+
+    unsigned char log_2_bucket_count;
+    res = bpf_probe_read(&log_2_bucket_count, sizeof(log_2_bucket_count), labels_map_ptr + 9);
+    if (res < 0)
+    {
+        return false;
+    }
+    u64 bucket_count = 1 << log_2_bucket_count;
+    void *label_buckets;
+    res = bpf_probe_read(&label_buckets, sizeof(label_buckets), labels_map_ptr + 16);
```

Comment:
Should we declare a map struct and collapse these 3 reads into one?   Would probably turn a 8 byte read, followed by 1 byte read, followed by 8 byte read into a memcpy I'm guessing?  Probably a wash performance wise but maybe smaller code?  Feel free to file under "thing for Tom to investigate".

Reply:
it might be possible to optimize, but the bpf stack is only 512bytes, which already caused me to have to move one of these to be stored in a per-cpu bpf map instead